### PR TITLE
Merge New CI Infrastructure Based on Trust/Cross Into Master

### PR DESCRIPTION
**Pull request by posborne**
See discussion on https://github.com/nix-rust/nix/pull/528.  This PR is to track the final decision of whether we are ready to merge the new CI infrastructure (and related bug fixes) back into master.  Initially, we know that the branch is not ready to merge (as there are still failing tests).

This PR acts as a single point of reference to get the current status of how close we are to merging things back into master.  A number of issues with "TRUST CI" in the title have been created for the work that appears to be required to get us to the point where this branch can be merged.

### CODE REVIEW

**Comment by Susurrus** [comment]
@berkowski @posborne I'd really like to get this merged and active for nix ASAP. What're the next steps for this work as I haven't been following it.

From my point of view, I think as long as this isn't a regression from the previous CI (as in the original platforms still pass) I'd like to get this merged. Once it's in-tree we can start to fix the other platforms as we go.

**Comment by berkowski** [comment]
#553 seems to solve all the compile-time issues for the architectures covered by trust but it's been pending a final review for the last 3 weeks.

Once that gets merged then the whole new-ci/master branch will need to be updated to the current working master, fixing any issues along the way.

**Comment by Susurrus** [comment]
@berkowski Excellent, then I'll take a look at #553 either today or tomorrow and we'll get that merged soon then (I'm a maintainer now). And then we can take another crack at this.

**Comment by berkowski** [comment]
Excellent (and congrats!)

**Comment by Susurrus** [comment]
One thing that I'd like to see added to this PR is an addition to the README describing which platforms `nix` officially supports as Tier 2 architectures (defined as only compiled-tested) and Tier 1 architectures (defined as compile- and run-tested). If all platforms have tests run on them, there's no point in distinguishing between Tier 1 and Tier 2, but let's still document the supported platforms. This will also help me when reviewing these changes so I know what should come out of this work.

File: .travis.yml [hunk 1]
```diff
@@ -1,102 +1,106 @@
...
-    - TRAVIS_CARGO_NIGHTLY_FEATURE=""
-  matrix:
-    - ARCH=x86_64
-    - ARCH=i686
+    - CRATE_NAME=nix
 
-os:  # OSX included in build matrix explicitly
-  - linux
+    # default job
+    - TARGET=x86_64-unknown-linux-gnu
```
Comment (Susurrus) [inline .travis.yml]: This needs an `env:` after the `-` so that it actually sets the environment variable in a way that Travis understands.

**Comment by asomers** [comment]
@susurrus FreeBSD and NetBSD aren't tested because Travis doesn't support those operating systems.  TRUST builds for them by using a cross-compiler, but it can't cross-test.

**Comment by Susurrus** [comment]
@asomers Thanks, that makes sense. We should document that in the README by listing them as Tier 2 platforms and also possibly in the `.travis.yml` file, though I'm less sure about that one.

@posborne Are you still the lead on this or are you otherwise occupied for a while?

File: .travis.yml (hunk 1 above)
Comment (Susurrus) [inline .travis.yml]: Also, I think this should be done within the `matrix` section. If you look at the Travis output, you see `TARGET` being declared twice.

**Comment by Susurrus** [comment]
As a summary for everyone, we currently have the following outstanding issues with this PR:
 - [ ] Fix `SIGABRT` for `i686-linux-musl`
 - [ ] Fix `SIGSEGV` for `x86_64-unknown-linux-musl`
 - [ ] Fix Travis issue for `i686-apple-darwin`
 - [ ] Fix long-running `sys::test_aio::test_lio_listio_signal` test for `mips-unknown-linux-gnu`
 - [ ] Fix long-running `sys::test_aio::test_lio_listio_signal` test for `mipsel-unknown-linux-gnu`
 - [ ] Fix `sys::test_select::test_select` for `powerpc-unknown-linux-gnu`

And I just realized this, but instead of targeting current stable Rust, we should target 1.9, which is our minimally supported version. Maybe then instead of the two nightly platforms we have, we would test stable. Nightly can give spurious failures, but it's of value to the compiler team, so I'm not so sure about this one.

I'd propose that we only address the `i686-apple-darwin` and `powerpc-unknown-linux-gnu` target failures as they should be easy and I think we should support those platforms. I'd then propose we disable build running tests on the other platforms for now so that we're all-green and open tracking issues for adding Tier 1 support for those untested platforms.

Since this testing is already more comprehensive than the current testing we get a step forward and this will allow us to move forward with other PRs that are basically blocking on this more comprehensive testing. Thoughts anyone?

**Comment by asomers** [comment]
Aren't the sys::test_aio::test_lio_listio_signal failures fixed by  https://github.com/posborne/rust-nix/commit/0e3ef36c7406f92483bfc75a37a382a1509fdb54 ?  Do you have that change in your branch?

**Comment by Susurrus** [comment]
@asomers Maybe they did, I've cherry-picked that commit and fixed the `TARGET` value for `x86_64-unknown-linux-gnu` to move this along a bit, so we'll see what Travis says...

**Comment by Susurrus** [comment]
For posterity looks like that commit was being tracked in #539, whoops.

**Comment by berkowski** [comment]
Remind me what the `i686-apple-darwin` failure is?  Seems like it passed in the last round.

I'd also recommend merging/rebasing the travis branch off the current master head before too much work is done.  A lot of other PR's have been merged into the main branch since the CI work started.

**Comment by Susurrus** [comment]
@berkowski There were no mac failures last time, so it looks like those are just intermittent failures. We'll keep an eye on them. I've rebased as you suggested so we'll see how this run goes.

**Comment by Susurrus** [comment]
How does everyone feel about altering the test platforms to all be rust 1.7 and keeping a stable and nightly for both 64-bit mac and linux?

**Comment by berkowski** [comment]
Sounds fine to me if 1.7 is the minimum supported version for `nix`.

I'd vote for adding beta for x64 linux/mac as well.  Nightly might throw up a lot of errors that 'fix' themselves, but an error showing up in beta will probably fall through to stable without action.

File: .travis.yml (hunk 1 above)
Comment (Susurrus) [inline .travis.yml]: Nevermind, found out this is a bug in Travis CI that they haven't put much effort into fixing: https://github.com/travis-ci/travis-ci/issues/4681

**Comment by Susurrus** [comment]
@berkowski Any ideas what the failures might be from those 5 failing cases? We should be able to easily test the `-musl` targets on any Linux machine, yes?

And switching to Rust 1.7.0 causes pretty much everything to fail because of what I think is a version incompatibility with `cross`. @berkowski I couldn't see a minimum supported rust/cargo version for `cross`, you know about that?

**Comment by berkowski** [comment]
"5 failures"... like from https://travis-ci.org/nix-rust/nix/builds/219888273 ?  The musl segfaults?

We're using `trust` v0.1.4 (see `ci/install.sh`).  I tried building it with rustc 1.7...

```
$cargo +1.7.0 build
   Compiling gcc v0.3.40
   Compiling cross v0.1.4 (file:///home/zac/repos/rust/cross)
   Compiling rustc_version v0.1.7
   Compiling kernel32-sys v0.2.2
   Compiling dbghelp-sys v0.2.0
build.rs:36:18: 36:19 error: expected one of `.`, `;`, or an operator, found `?`
build.rs:36         .output()?;
                             ^
error: aborting due to previous error
Build failed, waiting for other jobs to finish...
```

So seems like `dbghelp-sys` uses question syntax which didn't land in stable until `1.13.0` I think?

**Comment by Susurrus** [comment]
I guess the big question is are we okay with bumping the Rust minimum supported version to 1.13? I have a feeling we're going to see more and more libraries require 1.13 for the question-mark operator. Let's try this again with 1.13.

**Comment by Susurrus** [comment]
@berkowski Well it looks like everything works with 1.13, so I guess we'll set that as our new minimum version. And I added testing for beta. I'm not certain why nightly isn't showing up, I don't quite understand the syntax for allowing failures for some Travis CI builds.

@berkowski And yes, I'd like to resolve the musl segfaults. I'm betting the mips and powerpc failures are emulator related, but someone needs to investigate. Are you up for this work?

@posborne @berkowski Can either of you explain why there are 6 commented-out architectures for targetting? And also why aren't we at least compiling for android? I'm betting I'm missing where this conversation already happened, so I want to make sure I understand everything.

**Comment by berkowski** [comment]
I can try poking at the musl related things but I don't have any experience with that implementation and I'm pretty tight for free time for the next two weeks.  Hopefully it's something simple. 

As for the commented architectures: nix artificially refuses to complie on them.   I say artificially because there are #cfg clauses that import a nonexistent module named something like not_supported if the target arch doesn't match one of a list of "good" values.  libc has bindings for some of these arches,  maybe all, so I don't know why they're blocked.  I think they're fodder for future PRs.

**Comment by Susurrus** [comment]
Given that FreeBSD builds are broken on master and that we have to do some debugging work to fix the currently-failing 5 targets, I'm going to allow failures on those targets. If this passes, I'm going to merge this with master. I'd also like to do a new 0.9.0 release at this point since we've fixed a few targets (and an API break means we'll need to do a 0.9 release).

**Comment by asomers** [comment]
PR #579 might help with the mips failures.

**Comment by Susurrus** [comment]
Okay, everyone, I'm planning on merging this. I've segregated out failing targets into an `allowed_failure` section, got the CHANGELOG up to date, and updated the README with supported targets and the new minimum Rust version (1.13). I haven't heard any negative feedback on anything I've proposed so far so I'm going to assume that silence is a tacit agreement with this plan.

One thing that I should note is that I made it more explicit in the README about how well we support the various architectures. I split them into 3 tiers, in a similar vein to how Rust has 3 tiers of platform support.

**Comment by Susurrus** [comment]
@homu r+

**Comment by berkowski** [comment]
Nice work so far and good thought making it explicit in the readme what degrees of support are expected for each platform.

**Comment by Susurrus** [comment]
Okay, let's see if this will work this time:

@homu r+

**Comment by Susurrus** [comment]
@kamalmarhubi @posborne @fiveop Homu doesn't seem to be doing anything, but I'm not really familiar with Homu to debug this. Can any of you guys comment on why it's not working?

**Comment by kamalmarhubi** [comment]
Weird. Let me try!

@homu r+

**Comment by kamalmarhubi** [comment]
Hrm. That's upsetting.

**Comment by kamalmarhubi** [comment]
@Susurrus I'm not sure what's up. As far as I know it should work. At the same time, Homu has a database error, and has done for a while: http://homu.io/l/

**Comment by kamalmarhubi** [comment]
I will see if we can switch to [bors-ng](https://github.com/bors-ng/bors-ng), which has a hosted version as well.

**Comment by Susurrus** [comment]
@kamalmarhubi There's a `synchronize` button on the homu website and it asked for permissions to my account and nix-rust on GitHub. Do I need to do that so that I recognizes me or something?

**Comment by kamalmarhubi** [comment]
@Susurrus I just opened #580 to figure this out. You could try that, it sounds somewhat reasonable!

**Comment by posborne** [comment]
🎉 🎉  I've been off the radar for a little while helping land a big change (with a bunch of Rust) at work and have been doing a very poor job of maintaining.  Glad to see this land.  Thank you @asomers @Susurrus @kamalmarhubi and anyone else who contributed for getting this across the finish line!

**Comment by kamalmarhubi** [comment]
@berkowski too! This is a heroic effort, thanks to all of you! All I did was tell @homu to merge it. :-)

**Pull request merged**